protected/tests/unit/DatasetLogTest.php

```diff
@@ -0,0 +1,24 @@
+<?php
+
+
+class DatasetLogTest extends CDbTestCase
+{
+    public function testCreateDatasetLogEntry()
+    {
+        $datasetId = 8;
+        $fileName = "Pygoscelis_adeliae.scaf.fa.gz: file attribute deleted";
+        $fileModel = "FileAttribute";
+        $modelId = 11020;
+        $fileId = "/adminFile/update/id/664";
+        $datasetlog = new DatasetLog(); //Instantiate a new Object
+        $datasetlog->createDatasetLogEntry($datasetId, $fileName, $fileModel, $modelId, $fileId);
+        $this->assertNotNull($datasetlog);
+        $this->assertTrue(is_a($datasetlog, DatasetLog::class));
+//        $this->assertEquals($datasetId, $this->dataset_id);
+//        $this->assertEquals($fileName, $datasetlog->message);
+//        $this->assertEquals($fileModel, $datasetlog->model);
+//        $this->assertEquals($modelId, $datasetlog->model_id);
+//        $this->assertEquals($fileId, $datasetlog->url);
```

Comment:
Hi @rija I am going to pass this unit test for `createDatasetLogEntry()` function, do you know how to extract the value from `$datasetlog` to do the assertion? I tried a quite different ways, but  only `empty` value was returned.

Reply:
Hi @kencho51

Often, when testing is hard, it's a sign the design of the implementation is not optimal.

Here, use the factory approach like for CurationLog.

A factory method must return an instance of the class 
(It's like a real factory, whenever the production line is running, it outputs real base product, that can then be customised for specific markets further down the line)

Then you can make another method as part of the public API for DatasetLog (that's what ``createDatasetLogEntry()``seems to be)  that will make use of the factory method.

Also you must use specify return and parameter type definition in all  your functions' signature (I know I forgot to do it on the CurationLog public API methods). E.g:

```
public static function createDatasetLogEntry(int $id, string $fileName, object $fileModel, int $modelId, int $fileId): bool
```

Once you've separated instantiation and customisation and have articulated  the types of parameters and return value,  things (and what to test) will get a bit clearer.

Reply:
Hi @rija 
I have passed the acceptance test for the logging of addition and deletion in `features/dataset-log-file-attribute-add-delete.feature`, and I `assertTrue` for the return value for the function `createDatasetLogEntry()` in the unit test, do you think it is still needed to implement the `factory method`?

Reply:
Hi @kencho51,

Look at it this way:
As part of the same ticket, you've worked on CurationLog and DatasetLog whose usage mechanism are identical from a code perspective.
Their potential for changes also follow the same pattern (from somewhere in code we need to log info with parameters specific to a context).
How can we justify these two classes having differing implementation?

Often it's the right move to leave out refactoring for later when a new requirement comes up, 
but we should also maintain design consistency whilst we are working on something.

Some developers (most likely our future selves) who will one day stumble on DatasetLog, should go into CurationLog expecting to see the same design as a hint about their similarity and vice-versa, otherwise it's extra mind effort to figure why the difference.

Reply:
Hi @rija
I have implemented the factory pattern `datasetLogEntryFactory` for the function `createDatasetLogEntry`, and the `unit test` is passing.